python/ray/train/v2/_internal/execution/failure_handling/default.py

```diff
@@ -1,46 +1,85 @@
 import logging
 
-from ray.train import FailureConfig
-from ray.train.v2._internal.execution.failure_handling import (
-    FailureDecision,
-    FailurePolicy,
+from .failure_policy import FailureDecision, FailurePolicy
+from ray.train.v2._internal.exceptions import (
+    WorkerGroupStartupFailedError,
+    WorkerGroupStartupTimeoutError,
+)
+from ray.train.v2.api.config import FailureConfig
+from ray.train.v2.api.exceptions import (
+    ControllerError,
+    TrainingFailedError,
+    WorkerGroupError,
 )
-from ray.train.v2._internal.execution.worker_group import WorkerGroupPollStatus
 
 logger = logging.getLogger(__name__)
 
 
+RETRYABLE_CONTROLLER_ERRORS = (
+    WorkerGroupStartupFailedError,
+    WorkerGroupStartupTimeoutError,
+)
+
+
 class DefaultFailurePolicy(FailurePolicy):
     def __init__(self, failure_config: FailureConfig):
         super().__init__(failure_config)
-        self._total_failures = 0
+        self._worker_group_failures = 0
+        self._controller_failures = 0
+
+    def _log_decision(
+        self,
+        decision: FailureDecision,
+        training_failed_error: TrainingFailedError,
+        error_count: int,
+        retry_limit: int,
+    ):
+        logger.info(
+            f"[FailurePolicy] Decision: {decision}, "
+            f"Error count over retry limit: {error_count} / {retry_limit}, "
+            f"Error: {training_failed_error}"
+        )
+
+    def _is_retryable_error(self, training_failed_error: TrainingFailedError) -> bool:
+        if isinstance(training_failed_error, WorkerGroupError):
+            return True
+        elif isinstance(training_failed_error, ControllerError):
+            return isinstance(
+                training_failed_error.controller_failure, RETRYABLE_CONTROLLER_ERRORS
+            )
+        return False
 
     def make_decision(
-        self, worker_group_status: WorkerGroupPollStatus
+        self,
+        training_failed_error: TrainingFailedError,
     ) -> FailureDecision:
-        if not worker_group_status.errors:
-            return FailureDecision.NOOP
 
-        self._total_failures += 1
+        if not self._is_retryable_error(training_failed_error):
+            decision = FailureDecision.RAISE
+            error_count = 1
+            retry_limit = 0
+        else:
+            if isinstance(training_failed_error, ControllerError):
+                self._controller_failures += 1
+                error_count = self._controller_failures
+                retry_limit = (
+                    self.failure_config.controller_failure_limit
+                    if self.failure_config.controller_failure_limit != -1
+                    else float("inf")
+                )
+            else:
```

Comment:
```suggestion
            elif isinstance(training_failed_error, WorkerGroupError):
```

+ add a ValueError in the else, just for clarity